### PR TITLE
[SPARK-16077] [PYSPARK] catch the exception from pickle.whichmodule()

### PR DESCRIPTION
**Pull request by davies**
## What changes were proposed in this pull request?

In the case that we don't know which module a object came from, will call pickle.whichmodule() to go throught all the loaded modules to find the object, which could fail because some modules, for example, six, see https://bitbucket.org/gutworth/six/issues/63/importing-six-breaks-pickling

We should ignore the exception here, use `__main__` as the module name (it means we can't find the module).
## How was this patch tested?

Manual tested. Can't have a unit test for this.

### CODE REVIEW

**Comment by davies** [comment]
cc @JoshRosen

File: python/pyspark/cloudpickle.py [hunk 1]
```diff
@@ -169,7 +169,10 @@ def save_function(self, obj, name=None):
 
         if name is None:
             name = obj.__name__
-        modname = pickle.whichmodule(obj, name)
+        try:
+            modname = pickle.whichmodule(obj, name)
+        except Exception:
```
Comment (JoshRosen) [inline python/pyspark/cloudpickle.py]: Is there a narrower exception that we can catch (`ImportError` perhaps)?

File: python/pyspark/cloudpickle.py [hunk 2]
```diff
@@ -326,7 +329,10 @@ def save_global(self, obj, name=None, pack=struct.pack):
 
         modname = getattr(obj, "__module__", None)
         if modname is None:
-            modname = pickle.whichmodule(obj, name)
+            try:
+                modname = pickle.whichmodule(obj, name)
+            except Exception:
+                modname = '__main__'
```
Comment (JoshRosen) [inline python/pyspark/cloudpickle.py]: The use of `__main__` here seems consistent with the error-handling path on line 181 👍

**Comment by JoshRosen** [comment]
Do you mind leaving a comment in the code referencing this issue / maybe linking to that `six` issue so that it's easier for a reader to figure out what's going on? Otherwise, this looks good to me.

File: python/pyspark/cloudpickle.py (hunk 1 above)
Comment (davies) [inline python/pyspark/cloudpickle.py]: This will run into arbitary code, so anything could happen

**Comment by davies** [comment]
Merged into 1.6, 2.0 and master

**Pull request closed without merging**